client/lib/track-element-size/index.ts

```diff
@@ -0,0 +1,118 @@
+/**
+ * External dependencies
+ */
+import { useRef, useEffect, useState, useCallback } from 'react';
+import { throttle } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import afterLayoutFlush from 'lib/after-layout-flush';
+
+const THROTTLE_RATE = 200;
+
+// Indexable version of DOMRect / ClientRect.
+interface UseWindowResizeRect {
+	[index: string]: number | undefined;
+	top: number;
+	right: number;
+	bottom: number;
+	left: number;
+	width: number;
+	height: number;
+	x?: number;
+	y?: number;
+}
+
+// Function type for the expected callback for `useWindowResizeCallback`.
+interface UseWindowResizeCallback {
+	( node: Element | null | undefined ): void;
+}
+
+/**
+ * React hook that subscribes a consumer to changes to the bounding client rect of an element, based
+ * on window resize events.
+ * Does not notify when an element resizes due to reasons other than the window resizing.
+ * Uses throttling on the events, to avoid making changes too often.
+ *
+ * @param callback The function to call back on changes. Takes a single parameter: `boundingClientRect`.
+ *
+ * @returns The ref to be set on the consumer component.
+ */
+export function useWindowResizeCallback(
+	callback: ( boundingClientRect: UseWindowResizeRect ) => any
+) {
+	const lastRect = useRef< UseWindowResizeRect | null >( null );
+	const [ element, setElement ] = useState< Element | null >( null );
+
+	const callbackRef: UseWindowResizeCallback = useCallback( node => {
+		if ( node ) {
+			setElement( node );
+		}
+	}, [] );
+
+	useEffect( () => {
+		// Notify consumer of bounding client rect change.
+		const handleRectChange = ( rect: DOMRect | ClientRect ) => {
+			if ( rect ) {
+				const jsonRect = 'toJSON' in rect ? rect.toJSON() : rect;
+				// Avoid notifying consumer if nothing's changed.
+				for ( const property of Object.keys( jsonRect ) ) {
+					if (
+						lastRect.current === null ||
+						lastRect.current[ property ] !== jsonRect[ property ]
```

Comment:
This will perform some redundant checks, won't it? Is it enough to satisfy that a common, explicit set of properties satisfied by both `ClientRect` and `DomRect` are equal?

This prevents the need to make an indexable interface to a type that `DomRect` and `ClientRect` can satisfy.

```ts
		// Notify consumer of bounding client rect change.
		const handleRectChange = ( rect: DOMRect | ClientRect ) => {
			if ( rect ) {
				const jsonRect = 'toJSON' in rect ? rect.toJSON() : rect;
				// Avoid notifying consumer if nothing's changed.
				if (
					lastRect.current === null ||
					lastRect.current.bottom !== jsonRect.bottom ||
					lastRect.current.left !== jsonRect.left ||
					lastRect.current.right !== jsonRect.right ||
					lastRect.current.top !== jsonRect.top
				) {
					lastRect.current = jsonRect;
					return callback( jsonRect );
				}
			}
		};
```

Then we can use this type:

```ts
type UseWindowResizeRect = ReturnType< typeof Element.prototype.getBoundingClientRect >;
```

Reply:
I think that despite the verbosity, this is overall better, yes. Thanks for the suggestion!

Reply:
TypeScript has a builtin `ClientRect` type that's exactly the thing returned by `getBoundingClientRect`. Only for historical reasons (ancient IEs and TypeScript originate at the same company 🙂) it's not called `DOMRect`. Can we just use `ClientRect`?

Reply:
The type is `getBoundingClientRect(): ClientRect | DOMRect;`, which is annoying but also correct. Different browsers return different things, ie11 returns `ClientRect`, more modern browsers return `DOMRect`.

We should be able to say we return `ClientRect` because that is a minimal type that should be satisfied, but what would the benefit be?